// foxdriver/build/models/pagestyle.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _actor = require('../actor');

var _actor2 = _interopRequireDefault(_actor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Pagestyle extends _actor2.default {
    getComputed(node, markMatched, onlyMatched, filter) {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { computed } = yield _this.request('getComputed', {
                node, markMatched, onlyMatched, filter
            });
            return computed;
        })();
    }

    getAllUsedFontFaces(includePreviews, previewText, previewFontSize, previewFillStyle) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { fontFaces } = yield _this2.request('getAllUsedFontFaces', {
                includePreviews, previewText, previewFontSize, previewFillStyle
            });
            return fontFaces;
        })();
    }

    getUsedFontFaces(node, includePreviews, previewText, previewFontSize, previewFillStyle) {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { fontFaces } = yield _this3.request('getUsedFontFaces', {
                node, includePreviews, previewText, previewFontSize, previewFillStyle
            });
            return fontFaces;
        })();
    }

    getMatchedSelectors(node, property, filter) {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { rules, sheets, matched } = yield _this4.request('getMatchedSelectors', {
                node, property, filter
            });
            return { rules, sheets, matched };
        })();
    }

    getApplied(node, inherited, matchedSelectors, skipPseudo, filter) {
        return this.request('getApplied', {
            node, inherited, matchedSelectors, skipPseudo, filter
        });
    }

    isPositionEditable(node) {
        var _this5 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { value } = yield _this5.request('isPositionEditable', { node });
            return value;
        })();
    }

    getLayout(node, autoMargins) {
        return this.request('getLayout', { node, autoMargins });
    }

    addNewRule(node, pseudoClasses, editAuthored) {
        return this.request('addNewRule', { node, pseudoClasses, editAuthored });
    }
}
exports.default = Pagestyle;
module.exports = exports['default'];
