// foxdriver/build/domains/console.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _actor = require('../actor');

var _actor2 = _interopRequireDefault(_actor);

var _utils = require('../utils');

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Console extends _actor2.default {

    /**
     * Start the given Web Console listeners.
     *
     * @param  {String[]} listeners  listeners to listen to (default: 'PageError', 'ConsoleAPI')
     * @return {Promise.<Object>}    request response
     */
    startListeners(listeners = Console.listeners) {
        this.isEnabled = true;
        return this.request('startListeners', { listeners });
    }

    /**
     * Stop the given Web Console listeners.
     *
     * @param  {String[]} listeners  listeners to stop listen to (default: 'PageError', 'ConsoleAPI')
     * @return {Promise.<Object>}    request response
     */
    stopListeners(listeners = Console.listeners) {
        this.isEnabled = false;
        return this.request('stopListeners', { listeners });
    }

    /**
     * Retrieve the cached messages from the server.
     *
     * @param  {String[]}  messageTypes  type of cached message to get (default: 'PageError', 'ConsoleAPI')
     * @return {Promise.<Object>[]}      list of console messages
     */
    getCachedMessages(messageTypes = Console.listeners) {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            if (!_this.isEnabled) {
                throw new Error('Console is not enabled');
            }

            const resp = yield _this.request('getCachedMessages', { messageTypes });
            return resp.messages;
        })();
    }

    /**
     * Clear the cache of messages (console API calls only).
     *
     * @return {Promise.<Object>}  request response
     */
    clearMessagesCache() {
        return this.request('clearMessagesCache');
    }

    /**
     * Evaluate a JavaScript expression.
     *
     * @param  {String|Function}  script  js code to evaluate
     * @param  {Object}           args    arguments to pass to the function
     * @return {Promise.<Object, Error>}  result of the js function or an exception if script fails
     */
    evaluateJS(script, ...args) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const argsTransformed = (0, _utils.transformEvaluateArgs)(args);
            const text = (0, _utils.transformEvaluateScript)(script, argsTransformed);
            const { result, exception, exceptionMessage } = yield _this2.request('evaluateJS', { text });

            if (exception) {
                throw new Error(exceptionMessage);
            }

            return result;
        })();
    }

    /**
     * Get Web Console-related preferences on the server.
     *
     * @param  {String[]} preferences  An array with the preferences you want to retrieve.
     * @return {Promise.<Object>[]}    List of preferences
     */
    getPreferences(preferences) {
        return this.request('getPreferences', { preferences });
    }

    /**
     * Set Web Console-related preferences on the server.
     *
     * @param {Object} preferences  An object with the preferences you want to change.
     * @return {Promise.<Object>}   request response
     */
    setPreferences(preferences) {
        return this.request('setPreferences', { preferences });
    }

    /**
     * Autocomplete a JavaScript expression.
     *
     * @param  {String} text      The code you want to autocomplete.
     * @param  {Number} cursor    Cursor location inside the string. Index starts from 0.
     * @return {Promise.<Object>} request response
     */
    autocomplete(text, cursor) {
        return this.request('autocomplete', { text, cursor });
    }

    /**
     * Inspect the properties of an object.
     *
     * @return {Promise.<Object>} request response
     */
    inspectObjectProperties() {
        return this.request('inspectProperties');
    }
}
exports.default = Console;
Console.listeners = ['PageError', 'ConsoleAPI'];
module.exports = exports['default'];
