// foxdriver/build/domains/emulation.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _actor = require('../actor');

var _actor2 = _interopRequireDefault(_actor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This actor overrides various browser features to simulate different environments to
 * test how pages perform under various conditions.
 *
 * The design below, which saves the previous value of each property before setting, is
 * needed because it's possible to have multiple copies of this actor for a single page.
 * When some instance of this actor changes a property, we want it to be able to restore
 * that property to the way it was found before the change.
 *
 * A subtle aspect of the code below is that all get* methods must return non-undefined
 * values, so that the absence of a previous value can be distinguished from the value for
 * "no override" for each of the properties.
 */
class Emulation extends _actor2.default {
    /**
     * overwrite DPPX value
     *
     * @param  {Nimber}  dppx     dppx value
     * @return {Promise.Boolean}  true if value has changed
     */
    setDPPXOverride(dppx) {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this.request('setDPPXOverride', { dppx });
            return valueChanged;
        })();
    }

    /**
     * get current DPPX value
     *
     * @return {Promise.Number}  current dppx value
     */
    getDPPXOverride() {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { dppx } = yield _this2.request('getDPPXOverride');
            return dppx;
        })();
    }

    /**
     * clear DPPX value
     *
     * @return {Promise.Boolean}  true if value has changed
     */
    clearDPPXOverride() {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this3.request('clearDPPXOverride');
            return valueChanged;
        })();
    }

    /**
     * Transform the RDP format into the internal format and then set network throttling.
     *
     * @param  {Number}  downloadThroughput  throughput in byte/s
     * @param  {Number}  uploadThroughput    throughput in byte/s
     * @param  {Number}  latency             latency time in ms
     * @return {Promise.Boolean}             true if value has changed
     */
    setNetworkThrottling(downloadThroughput, uploadThroughput, latency) {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this4.request('setNetworkThrottling', {
                options: {
                    downloadThroughput,
                    uploadThroughput,
                    latency
                }
            });
            return valueChanged;
        })();
    }

    /**
     * Get network throttling and then transform the internal format into the RDP format.
     *
     * @return {Promise.<Object>}  state of current throttle
     */
    getNetworkThrottling() {
        var _this5 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { state } = yield _this5.request('getNetworkThrottling');
            return state;
        })();
    }

    /**
     * clear network throttling
     *
     * @return {Promise.Boolean}   true if value has changed
     */
    clearNetworkThrottling() {
        var _this6 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this6.request('clearNetworkThrottling');
            return valueChanged;
        })();
    }

    /**
     * overwrite touch events
     *
     * @param  {Boolean}  flag    true if overwrite is enabled
     * @return {Promise.Boolean}  true if value has changed
     */
    setTouchEventsOverride(flag) {
        var _this7 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this7.request('setTouchEventsOverride', { flag });
            return valueChanged;
        })();
    }

    /**
     * check if touch event overwrite is enabled
     *
     * @return {Promise.Boolean}  true if enabled
     */
    getTouchEventsOverride() {
        var _this8 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { flag } = yield _this8.request('getTouchEventsOverride');
            return flag;
        })();
    }

    /**
     * clear state of touch event overwrite
     *
     * @return {Promise.Boolean}  true if enabled
     */
    clearTouchEventsOverride() {
        var _this9 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this9.request('clearTouchEventsOverride');
            return valueChanged;
        })();
    }

    /**
     * Overwrite user agent
     *
     * @param  {String}  userAgent  new user agent
     * @return {Promise.Boolean}    true if value has changed
     */
    setUserAgentOverride(userAgent) {
        var _this10 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this10.request('setUserAgentOverride', { flag: userAgent });
            return valueChanged;
        })();
    }

    /**
     * Get current user agent overwrite
     *
     * @return {Promise.String}  current user agent
     */
    getUserAgentOverride() {
        var _this11 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { userAgent } = yield _this11.request('getUserAgentOverride');
            return userAgent;
        })();
    }

    /**
     * Clear user agent overwrite
     *
     * @return {Promise.Boolean}    true if value has changed
     */
    clearUserAgentOverride() {
        var _this12 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { valueChanged } = yield _this12.request('clearUserAgentOverride');
            return valueChanged;
        })();
    }
}
exports.default = Emulation;
module.exports = exports['default'];
