// foxdriver/build/domains/performance.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _actor = require('../actor');

var _actor2 = _interopRequireDefault(_actor);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * This actor wraps the Performance module at devtools/shared/shared/performance.js
 * and provides RDP definitions.
 *
 * @see devtools/shared/shared/performance.js for documentation.
 */
class Performance extends _actor2.default {
    connect(options) {
        return this.request('connect', { options });
    }

    /**
     * Checks whether or not a new recording is supported by the PerformanceFront.
     *
     * @return {Promise.Boolean}  true if it can record performance
     */
    canCurrentlyRecord(options) {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { value } = yield _this.request('canCurrentlyRecord');
            return value;
        })();
    }

    /**
     * Begins a recording session.
     *
     * @param  {Boolean}  options.withMarkers
     *                    include markers
     * @param  {Boolean}  options.withTicks
     *                    include ticks
     * @param  {Boolean}  options.withMemory
     *                    include memory
     * @param  {Boolean}  options.withAllocations
     *                    include allocations
     * @param  {Boolean}  options.allocationsSampleProbability
     *                    include allocation sample probability
     * @param  {Boolean}  options.allocationsMaxLogLength
     *                    include allocations max log length
     * @param  {Boolean}  options.bufferSize
     *                    include buffer size
     * @param  {Boolean}  options.sampleFrequency
     *                    include sample frequency
     * @param  {Boolean}  options.console
     *                    include console
     * @param  {String}   options.label
     *                    label of session recording
     * @param  {Boolean}  options.realtimeMarkers
     *                    use real time markers
     *
     * @return {Promise.<Object>} A promise that is resolved once recording has started.
     */
    startRecording(options) {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { recording } = yield _this2.request('startRecording', { options });
            return recording;
        })();
    }

    /**
     * Manually ends the recording session for the corresponding PerformanceRecording.
     *
     * @param {PerformanceRecording} model
     *        The corresponding PerformanceRecording that belongs to the recording
     *        session wished to stop.
     * @return {Promise.PerformanceRecording}
     *         Returns the same model, populated with the profiling data.
     */
    stopRecording(options) {
        var _this3 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { recording } = yield _this3.request('stopRecording', { options });
            return recording;
        })();
    }

    /**
     * Checks all currently stored recording handles and returns a boolean
     * if there is a session currently being recorded.
     *
     * @return {Boolean} true if actor is currently recording
     */
    isRecording() {
        var _this4 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { isRecording } = yield _this4.request('isRecording');
            return isRecording;
        })();
    }

    /**
     * Returns all current recordings.
     * @return {Promise.PerformanceRecording}  returns model with populated profing data
     */
    getRecordings() {
        var _this5 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { recordings } = yield _this5.request('getRecordings');
            return recordings;
        })();
    }

    /**
     * Returns the configurations set on underlying components, used in tests.
     * Returns an object with `probability`, `maxLogLength` for allocations, and
     * `features`, `threadFilters`, `entries` and `interval` for profiler.
     *
     * @return {Promise.Object}
     */
    getConfiguration() {
        var _this6 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const { config } = yield _this6.request('getConfiguration');
            return config;
        })();
    }
}
exports.default = Performance;
module.exports = exports['default'];
