// foxdriver/build/browser.js
'use strict';

Object.defineProperty(exports, "__esModule", {
    value: true
});

var _asyncToGenerator2 = require('babel-runtime/helpers/asyncToGenerator');

var _asyncToGenerator3 = _interopRequireDefault(_asyncToGenerator2);

var _actor = require('./actor');

var _actor2 = _interopRequireDefault(_actor);

var _client = require('./client');

var _client2 = _interopRequireDefault(_client);

var _tab = require('./tab');

var _tab2 = _interopRequireDefault(_tab);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

class Browser extends _actor2.default {
    constructor(host, port) {
        const client = new _client2.default(host, port);

        client.on('error', error => this.emit('error', error));
        client.on('end', () => this.emit('end'));
        client.on('timeout', () => this.emit('timeout'));

        super(client, 'root');
    }

    get preference() {
        return this._get('preference');
    }
    get actorRegistry() {
        return this._get('actorRegistry');
    }
    get addons() {
        return this._get('addons');
    }
    get device() {
        return this._get('device');
    }
    get heapSnapshotFile() {
        return this._get('heapSnapshotFile');
    }

    connect() {
        var _this = this;

        return (0, _asyncToGenerator3.default)(function* () {
            yield _this.client.connect();
            _this.client.expectReply(_this.name, function ({ traits }) {
                _this.traits = traits;
            });
            _this.tabs = yield _this.listTabs();
            return _this.tabs;
        })();
    }

    disconnect() {
        this.client.disconnect();
    }

    close() {
        this.disconnect();

        /**
         * only shut down browser if started via launcher
         */
        if (!this.firefoxProcess) {
            return console.error('Can\'t close the browser because client was attached to an' + 'already opened Firefox instance');
        }

        this.firefoxProcess.kill();
    }

    listTabs() {
        var _this2 = this;

        return (0, _asyncToGenerator3.default)(function* () {
            const listTabsResponse = yield _this2.request('listTabs');
            _this2.setActors(listTabsResponse);
            return listTabsResponse.tabs.map(function (tab) {
                return new _tab2.default(_this2.client, tab.actor, tab);
            });
        })();
    }
}
exports.default = Browser;
module.exports = exports['default'];
